src/webgpu/api/validation/attachment_compatibility.spec.ts

```diff
@@ -0,0 +1,305 @@
+import { range } from '../../../common/framework/util/util.js';
+import {
+  kRegularTextureFormatInfo,
+  kRegularTextureFormats,
+  kSizedDepthStencilFormats,
+  kUnsizedDepthStencilFormats,
+  RegularTextureFormat,
+} from '../../capability_info.js';
+
+import { ValidationTest, CommandBufferMaker } from './validation_test.js';
+
+// TODO: Update maximum color attachments when defined
+const kColorAttachmentCounts = range(4, i => i + 1);
+const kDepthStencilAttachmentFormats = [
+  undefined,
+  ...kSizedDepthStencilFormats,
+  ...kUnsizedDepthStencilFormats,
+] as const;
+
+class F extends ValidationTest {
+  createAttachmentTextureView(format: GPUTextureFormat) {
+    return this.device
+      .createTexture({
+        size: [1, 1, 1],
+        format,
+        usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+      })
+      .createView();
+  }
+
+  createColorAttachment(format: GPUTextureFormat): GPURenderPassColorAttachmentDescriptor {
+    return {
+      attachment: this.createAttachmentTextureView(format),
+      loadValue: [0, 0, 0, 0],
+    };
+  }
+
+  createDepthAttachment(format: GPUTextureFormat): GPURenderPassDepthStencilAttachmentDescriptor {
+    return {
+      attachment: this.createAttachmentTextureView(format),
+      depthLoadValue: 0,
+      depthStoreOp: 'clear',
+      stencilLoadValue: 1,
+      stencilStoreOp: 'clear',
+    };
+  }
+
+  createPassOrBundleEncoder(
+    encoderType: 'render pass' | 'render bundle',
+    colorFormats: Iterable<GPUTextureFormat>,
+    depthStencilFormat?: GPUTextureFormat
+  ): CommandBufferMaker<GPURenderPassEncoder | GPURenderBundleEncoder> {
+    const encoder = this.device.createCommandEncoder();
+    const passDesc: GPURenderPassDescriptor = {
+      colorAttachments: Array.from(colorFormats, format => this.createColorAttachment(format)),
+      depthStencilAttachment:
+        depthStencilFormat !== undefined
+          ? this.createDepthAttachment(depthStencilFormat)
+          : undefined,
+    };
+    const pass = encoder.beginRenderPass(passDesc);
+    switch (encoderType) {
+      case 'render bundle': {
+        const bundleEncoder = this.device.createRenderBundleEncoder({
+          colorFormats,
+          depthStencilFormat,
+        });
+
+        return {
+          encoder: bundleEncoder,
+          finish() {
+            const bundle = bundleEncoder.finish();
+            pass.executeBundles([bundle]);
+            pass.endPass();
+            return encoder.finish();
+          },
+        };
+      }
+      case 'render pass':
+        return {
+          encoder: pass,
+          finish() {
+            pass.endPass();
+            return encoder.finish();
+          },
+        };
+    }
+  }
+
+  createRenderPipeline(
+    colorStates: Iterable<GPUColorStateDescriptor>,
+    depthStencilState?: GPUDepthStencilStateDescriptor
+  ) {
+    const wgslVertex = `
+    [[builtin(position)]] var<out> position : vec4<f32>;
+
+    [[stage(vertex)]]
+    fn main() -> void {
```

Comment:
nit: we try to have a convention of putting the `[[stage]]` and `fn main` on the same line.